perfkitbenchmarker/linux_benchmarks/iperf_benchmark.py

```diff
@@ -152,9 +139,166 @@ def _RunIperf(sending_vm, receiving_vm, receiving_ip_address, thread_count,
       'sending_thread_count': thread_count,
       'sending_zone': sending_vm.zone,
       'runtime_in_seconds': FLAGS.iperf_runtime_in_seconds,
-      'ip_type': ip_type
+      'ip_type': ip_type,
   }
-  return sample.Sample('Throughput', total_throughput, 'Mbits/sec', metadata)
+
+  if protocol == TCP:
+
+    iperf_cmd = (f'iperf --enhancedreports --client {receiving_ip_address} --port {IPERF_PORT} '
+                 f'--format m --time {FLAGS.iperf_runtime_in_seconds} --parallel {thread_count}')
+
+    if FLAGS.iperf_tcp_per_stream_bandwidth:
+      iperf_cmd += f' --bandwidth {FLAGS.iperf_tcp_per_stream_bandwidth}M'
+
+    # the additional time on top of the iperf runtime is to account for the
+    # time it takes for the iperf process to start and exit
+    timeout_buffer = FLAGS.iperf_timeout or 30 + thread_count
+    stdout, _ = sending_vm.RemoteCommand(iperf_cmd, should_log=True,
+                                         timeout=FLAGS.iperf_runtime_in_seconds + timeout_buffer)
+
+    window_size_match = re.search(r'TCP window size: (?P<size>\d+\.?\d+) (?P<units>\S+)', stdout)
+    window_size = float(window_size_match.group('size'))
+    window_size_unit = window_size_match.group('units')
+
+    buffer_size = float(re.search(r'Write buffer size: (?P<buffer_size>\d+\.\d+) \S+', stdout).group('buffer_size'))
+
+    multi_thread = re.search((r'\[SUM\]\s+\d+\.\d+-\d+\.\d+\s\w+\s+(?P<transfer>\d+)\s\w+\s+(?P<throughput>\d+)'
+                              r'\s\w+/\w+\s+(?P<write>\d+)/(?P<err>\d+)\s+(?P<retry>\d+)\s*'),
+                             stdout)
+    # Iperf output is formatted differently when running with multiple threads vs a single thread
+    if multi_thread:
+      # Write, error, retry
+      write = int(multi_thread.group('write'))
+      err = int(multi_thread.group('err'))
+      retry = int(multi_thread.group('retry'))
+
+    # if single thread
+    else:
+      # Write, error, retry
+      match = re.search(r'\d+ Mbits/sec\s+(?P<write>\d+)/(?P<err>\d+)\s+(?P<retry>\d+)', stdout)
+      write = int(match.group('write'))
+      err = int(match.group('err'))
+      retry = int(match.group('retry'))
+
+    match = re.findall((r'\d+ Mbits/sec\s+ \d+/\d+\s+\d+\s+(?P<cwnd>-*\d+)(?P<cwnd_unit>\w+)/(?P<rtt>\d+)'
```

Comment:
Can use you re.finditer so that you can actually use the named capture groups in the code below?
https://stackoverflow.com/questions/11103856/re-findall-which-returns-a-dict-of-named-capturing-groups

Reply:
Switched to re.finditer